src/domain/core/src/services/query_service.rs

```diff
@@ -73,13 +76,29 @@ pub trait QueryService: Send + Sync {
     // number of files we collect to construct the dataframe.
     //
     /// Returns a [DataFrame] representing the contents of an entire dataset
-    async fn get_data(&self, dataset_ref: &DatasetRef) -> Result<DataFrame, QueryError>;
+    async fn get_data(&self, dataset_ref: &DatasetRef) -> Result<DataFrameWithContext, QueryError>;
 
     /// Lists engines known to the system and recommended for use
     async fn get_known_engines(&self) -> Result<Vec<EngineDesc>, InternalError>;
 }
 
-///////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////
+
+/// Dataframe and its context. Without context, we cannot perform operations on
+/// the dataframe
+#[derive(Clone)]
+pub struct DataFrameWithContext {
+    pub ctx: SessionContext,
```

Comment:
Is this because you make `KamuShema` hold a weak reference to state and need `SessionContext` to outlive it?

Reply:
`DataFrame` cannot be used without a context present -- previously dataframe could be used without explicit context because it leaked (never released)

Reply:
> Secondly, If we do have to pass SessionContext around I would prefer it to become a part of QueryResponse structure

It looks like I was trying to accomplish this with `DataFrameWithContext` structure, but apparently I didn't pick the wrong name.

Q: Does your comment relate more to naming, or more to the proposed composition?